### ISSUE TITLE
Refactor Click CLI to Rest API

### ISSUE THREAD

**Issue by johnding1996** [LionSR/TeXRA#1]
Without changing actural functionality, we want to refactor the CLI interface to APIs.

- [ ] Refactor backend's click CLIs to fastapis
- [ ] Refactor frontend's terminal CLI calls to rest API calls
- [ ] Integrated test of all routines

**Comment by johnding1996** [comment]
Relevant code of backend part all in, need similar feature of fastapi to handle lots of reused args in APIs.

https://github.com/LionSR/coauthor/blob/4062f0e89b2549eca254dc86116442d9c426b533/coauthor/cli.py#L42-L67

**Comment by LionSR** [comment]
for the frontend side two kinds of tasks 

- The main ``execute'' task:

https://github.com/LionSR/coauthor/blob/4062f0e89b2549eca254dc86116442d9c426b533/coauthor-for-vs-code/src/commands.ts#L315-L368

- The ``housekeeping'' tasks, including latexdiff/pack/clean etc

https://github.com/LionSR/coauthor/blob/4062f0e89b2549eca254dc86116442d9c426b533/coauthor-for-vs-code/src/commands.ts#L210-L244

not sure if this should be moved to the backend? Maybe it is better to just rewrite with typescript/javascript?

We might need to temporary keep this part in the python backend to avoid change two many things at the same time

**Comment by johnding1996** [comment]
yes we can keep the current click clis and `terminal.sendTest`, what is needed is a python module for fast apis and `https.post` things in parallel.

**Comment by LionSR** [comment]
The tricky bit might be that ideally we would like the backend(the server) to be ran on a remote VPS, but the returned output still need to be parsed, (split), and saved. This part might be better to be left with the frontend, which will query the backend if see if jobs are finished.  If finished, then the output is parsed (there are some tricky postprocessings though because the model do not reliable output latex and xml tags yet, this can happen in the backend)

Edit: The "returned output" here refers to the output obtained from the LLM model, and this output sometimes has imperfect XML, so it needs some special processing in the backend. The LLM might output \end{document} as </document> and similar issues. Then this error-free XML will be parsed and split into multiple LaTeX files.

In addition, Often the latexdiff should also be executed (this part should happen locally) to check the diffs. Now the logic is interwined (when the first document is finished, the latexdiff is executed before the reflection round started). We should make sure this still works. Now I think the better solution is to make these housekeeping logic to the frontend where IO operations are performed.

**Comment by LionSR** [comment]
fastapi has a CLI app sibling [](https://github.com/fastapi/fastapi?tab=readme-ov-file)
https://typer.tiangolo.com/

**Comment by johnding1996** [comment]
Prototyped changes with help from claude in #2

**Comment by LionSR** [comment]
after discussion:

TODOs:

- [ ] Check post messages to make sure front/backend matches @johnding1996 
- [ ] Send the Absolute Path/Directory to the Backend in POST as well @johnding1996 
- [ ] Read the Directory and Use the directory information in the backend @LionSR 

this would be a temp fix, to maintain the same behavior as before.

In addition, indent/pack/clean might be better to be handled in the frontend using sendterminals etc. Variables should be extracted correctly though.

**Comment by LionSR** [comment]
The better logic would still be to move the FileIO operations to the frontend (I think). See #6

**Comment by LionSR** [comment]
some progress signals still need to be streamed in the frontend, see #8
